### PR TITLE
(PUP-5831) Add binary type

### PR DESCRIPTION
**Pull request by hlindberg**
This adds a Binary type to the type system along with language operations. This PR also contains a function that reads a file in binary mode and returns the content as a Binary. PR also has exploratory work for setting the content of a File to be Binary, and to make this work for a puppet apply.

What remains to make the Binary type useful is the ability to include it in a catalog (i.e. basically serialization).

### CODE REVIEW

File: lib/puppet/functions/binary_file.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+Puppet::Functions.create_function(:binary_file, Puppet::Functions::InternalFunction) do
+  dispatch :binary_file do
+    scope_param
+    param 'String', :path
+  end
+
+  def binary_file(scope, unresolved_path)
+    path = nil
+    found = Puppet::Parser::Files.find_file(unresolved_path, scope.compiler.environment)
+    if found && Puppet::FileSystem.exist?(found)
```
Comment (thallgren) [inline lib/puppet/functions/binary_file.rb]: Unless I read the code incorrectly, a file returned from `Puppet::Parser::Files#find_file` is guaranteed to exist (see module.rb:137), so this extra `#exist?` is unnecessary and might be expensive.

File: lib/puppet/functions/find_file.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+  def find_file_array(scope, array)
+    find_file(scope, *array)
+  end
+
+  def find_file(scope, *args)
+    path = nil
+    args.each do |file|
+      found = Puppet::Parser::Files.find_file(file, scope.compiler.environment)
+      if found && Puppet::FileSystem.exist?(found)
```
Comment (thallgren) [inline lib/puppet/functions/find_file.rb]: Same here, the extra `#exists?()` is unnecessary.

File: lib/puppet/functions/binary_file.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+      path = found
+    end
+
+    if path
+      Puppet::Pops::Types::PBinaryType::Binary.from_binary_string(Puppet::FileSystem.binread(path))
```
Comment (thallgren) [inline lib/puppet/functions/binary_file.rb]: If we're keen on keeping the number of calls to `#stat` to a minimum, then we should have a `#find_file` variant that doesn't check for existence at all and then rely on the `Errno::ENOENT` raised here when the file is missing. At present, this call will result in a third `#stat` (or similar).

File: lib/puppet/functions/new.rb [hunk 4]
```diff
@@ -26,6 +26,6 @@ def new_function_for_type(t, scope)
   end
 
   def assert_type(type, value)
-    Puppet::Pops::Types::TypeAsserter.assert_instance_of('new():', type, value)
+    Puppet::Pops::Types::TypeAsserter.assert_instance_of("Converted value from #{type.to_s}.new()", type, value)
```
Comment (thallgren) [inline lib/puppet/functions/new.rb]: Better to use:
```
Puppet::Pops::Types::TypeAsserter.assert_instance_of(['Converted value from %s.new()', type], type, value)
```
to avoid unnecessary formatting when value is an instance of type.

File: lib/puppet/pops/evaluator/access_operator.rb [hunk 5]
```diff
@@ -30,6 +30,10 @@ def access_Object(o, scope, keys)
     fail(Issues::OPERATOR_NOT_APPLICABLE, @semantic.left_expr, :operator=>'[]', :left_value => o)
   end
 
+  def access_Binary(o, scope, keys)
+    Puppet::Pops::Types::PBinaryType::Binary.from_binary_string(access_String(o.binary_buffer, scope, keys))
```
Comment (thallgren) [inline lib/puppet/pops/evaluator/access_operator.rb]: How can I pick an individual byte from a Binary and read it as an Integer?

File: lib/puppet/pops/types/p_binary_type.rb [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+      Base64.encode64(@binary_buffer)
+    end
+
+    # Returns the binary content as a url safe base64 string (where + and / are replaced by - and _)
+    #
+    def urlsafe_to_s
+      Base64.urlsafe_encode64(@binary_buffer)
+    end
+
+    def eql?(o)
```
Comment (thallgren) [inline lib/puppet/pops/types/p_binary_type.rb]: Should probably have a #hash method as well.

File: lib/puppet/pops/types/p_binary_type.rb [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+    o.is_a?(Binary)
+  end
+
+  def eql?(o)
+    self.class == o.class
+  end
+
+  def hash?
+    super
+  end
```
Comment (thallgren) [inline lib/puppet/pops/types/p_binary_type.rb]: This method is redundant.

File: lib/puppet/pops/types/p_binary_type.rb [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+      @binary_buffer.length
+    end
+  end
+
+  def self.register_ptype(loader, ir)
+    create_ptype(loader, ir, 'AnyType')
+  end
+
+  def initialize()
+  end
```
Comment (thallgren) [inline lib/puppet/pops/types/p_binary_type.rb]: This method is redundant.

File: lib/puppet/pops/types/string_converter.rb [hunk 9]
```diff
@@ -385,6 +386,21 @@ def initialize
...
   # | p         | same as d
   #
+  # ### Binary (value)
+  #
+  # | Format    | Default formats
+  # | ------    | ---------------
+  # | s         | binary as unquoted characters
+  # | p         | 'Binary("<base64strict>")'
+  # | b         | '<base64>' - base64 string with newlines inserted
+  # | B         | '<base64strict>' - base64 strict string (without newlines inserted)
```
Comment (thallgren) [inline lib/puppet/pops/types/string_converter.rb]: It seems a bit inconsistent that '%s' maps to unquoted with newlines while '%B' maps to the `Binary#to_s()` method. Intuitively, I'd assume that '%s' would be the same as `#to_s()`.

File: lib/puppet/pops/types/p_binary_type.rb [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+
+  DEFAULT = PBinaryType.new
+
+  protected
+
+  def _assignable?(o, guard)
+    o.class == self.class
+  end
+
+end
```
Comment (thallgren) [inline lib/puppet/pops/types/p_binary_type.rb]: The ticket mentions encoding and how that affects assignability. Was that abandoned or deferred?

File: lib/puppet/functions/binary_file.rb (hunk 1 above)
Comment (hlindberg) [inline lib/puppet/functions/binary_file.rb]: While I don't think this matters performance wise, there is no need to check twice. Will remove the extra existence check.

File: lib/puppet/functions/binary_file.rb (hunk 3 above)
Comment (hlindberg) [inline lib/puppet/functions/binary_file.rb]: Sure. Given though that, there is an issue currently with ENOENT vs. permission failure on JRuby and that the expected frequency of calls to `binary_file` is both low, and minor in relation to actually reading the file, the optimization of this would be best performed later - and then looking at file reading and existence checks in general.

File: lib/puppet/pops/evaluator/access_operator.rb (hunk 5 above)
Comment (hlindberg) [inline lib/puppet/pops/evaluator/access_operator.rb]: Create an Array from the binary. Slice it before or after the conversion.

File: lib/puppet/pops/types/string_converter.rb (hunk 9 above)
Comment (hlindberg) [inline lib/puppet/pops/types/string_converter.rb]: We could use a different letter if that helps. The intent is that when you use %s you want the binary as text. Normally binary data would not be text and you would spew out binary gobbledygook. So the distinction needs to exist. The `to_s` variant is typically used blindly in all sorts of "lets print it out" so, it can not be the "binary as text" option.

File: lib/puppet/pops/types/p_binary_type.rb (hunk 10 above)
Comment (hlindberg) [inline lib/puppet/pops/types/p_binary_type.rb]: Abandoned. Bad idea. A Binary is simply a binary buffer and any encoding is about how it is presented in some medium.

File: lib/puppet/functions/binary_file.rb (hunk 1 above)
Comment (thallgren) [inline lib/puppet/functions/binary_file.rb]: What makes you think that it doesn't matter? I think that an `#exist?` call is likely to do a `stat()`, and even if the result is cached by the OS, it's still a switch between user space and system space.

**Comment by hlindberg** [comment]
I tried with removing what appears to be redundant existence checks, but that did not end well. It then errors with bad error message. I reverted that (local) change favouring getting a better error message.

**Comment by hlindberg** [comment]
@thallgren Do you think it is worth changing the %s format to something else? I have not made that change.

File: lib/puppet/functions/binary_file.rb (hunk 1 above)
Comment (thallgren) [inline lib/puppet/functions/binary_file.rb]: I see now that if an absolute path is used, the `#exists?` call is still needed since `#find_file` simply assumes that the absolute path exists. Guess that's another thing we could take a look at if/when we do a bigger overhaul to lessen the number of file accesses.

**Comment by thallgren** [comment]
The current interpretation of '%s' is  "binary as unquoted characters". This is broken since there are several character sets to choose from. What encoding was/is assumed? ISO-8859-1, extended ASCII, UTF-8, or something else? Representing a `Binary` as a `String` should never be possible without knowledge of encoding. In other words, the fact that Ruby handles this badly should not be allowed to leak into Puppet.

The String converter already defines the default encoding as strict base64 (it uses '%B' if nothing else is specified). That is a good choice. A default encoding when using '%s' must also be specified. The best candidate is probably 'UTF-8' and the documentation for '%s' should be changed to "binary UTF-8 encoded string".

What if a `Binary` that is known to represent an ISO-8859-1 encoded String? Should it be possible to convert that using the `StringConverter`? Could encoding be passed as a second argument somehow? Then again, that goes both ways so perhaps we need a separate ticket for encode/decode ?

**Comment by thallgren** [comment]
Please note that tests are failing for Ruby <= 2.0.0.

**Comment by hlindberg** [comment]
Will look at broken tests.
Agree, that %s should state unquoted UTF-8 encoded string (it should be an error if it is not valid UTF-8). It should also be more careful when creating the string.

**Comment by hlindberg** [comment]
I have addressed all comments, and I found a couple of problems myself:
* the 3.x stubs for the two new functions binary_file and find_file were missing in the PR (now added)
* the %s format now enforces UTF-8 encoding (the '%#s' format does not however as it outputs escaped bytes). It errors if strings is invalid UTF-8.
* the older rubies issue was that the base64 methods have bugs that simply truncates unpadded base64 encoded text - manual padding was added to fix this

**Comment by hlindberg** [comment]
When merging this also merge the corresponding specification PR

File: lib/puppet/parser/functions/new.rb [hunk 11]
```diff
@@ -626,6 +626,21 @@
...
+| ------    | ---------------
+| s         | binary as unquoted UTF-8 characters (errors if byte sequence is invalid UTF-8)
+| p         | 'Binary("<base64strict>")'
+| b         | '<base64>' - base64 string with newlines inserted
+| B         | '<base64strict>' - base64 strict string (without newlines inserted)
+| u         | '<base64urlsafe>' - base64 urlsafe string
+| t         | 'Binary' - outputs the name of the type only
+| T         | 'BINARY' - output the name of the type in all caps only
+
+The alternate form flag `#` will quote the binary or base64 text output
```
Comment (thallgren) [inline lib/puppet/parser/functions/new.rb]: A '%#s' doesn't just quote the string. It also uses escaped bytes rather than as UTF-8. I think that deserves a mention.

File: spec/unit/pops/types/p_binary_type_spec.rb [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+
+        it 'the Binary type is equal to itself only' do
+          code = <<-CODE
+            $t = Binary
+            notice(Binary =~ Type[ Binary ])
+            notice(Binary == Binary)
+            notice(Binary < Binary)
+            notice(Binary > Binary)
+          CODE
+          expect(eval_and_collect_notices(code)).to eql(['true', 'true', 'false', 'false'])
```
Comment (thallgren) [inline spec/unit/pops/types/p_binary_type_spec.rb]: The test doesn't assert what it claims to assert.

File: spec/unit/pops/types/p_binary_type_spec.rb [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+    it "has a length method in ruby returning the length measured in bytes" do
+      # \u{1f452} is "woman's hat emoji - 4 bytes in UTF-8"
+      a_binary = PBinaryType::Binary.new("\u{1f452}")
+      expect(a_binary.length).to be(4)
+    end
+  end
+
+end
+end
+end
```
Comment (thallgren) [inline spec/unit/pops/types/p_binary_type_spec.rb]: A new type should also be added to the `all_types` array defined in `spec/shared_contexts/types_setup.rb`. Doing so will test it together with other types in several different ways. It will also test that the type is serializable though, so perhaps such an addition must wait until Binary serialization has been added? Need to be remembered (this also applies to the new `Sensitive` type. It hasn't been added there either).

File: lib/puppet/parser/functions/new.rb (hunk 11 above)
Comment (hlindberg) [inline lib/puppet/parser/functions/new.rb]: ok, will fix

File: lib/puppet/parser/functions/new.rb [hunk 14]
```diff
@@ -810,7 +827,59 @@
...
+```
+
+The formats have the following meaning:
+
+| format | explanation |
+| ----   | ----        |
+| b | The data is in base64 encoding, padding as required by base64 strict is added by default
+| u | The data is in URL safe base64 encoding
+| B | The data is in base64 strict encoding
+| s | The data is a puppet string used verbatim as a binary byte sequence
```
Comment (hlindberg) [inline lib/puppet/parser/functions/new.rb]: well, unfortunately it is not specified in Puppet that strings in the language are UTF-8. The only thing we know is that strings created from text in a .pp file is UTF-8 since every .pp file is UTF-8. Strings returned from functions are however not required to have any particular encoding since they can represent binary file content or some string. It should really not be that way. I can change the wording of the explanation and say something like "used as a binary byte sequence with encoding ignored".

File: spec/unit/pops/types/p_binary_type_spec.rb (hunk 13 above)
Comment (hlindberg) [inline spec/unit/pops/types/p_binary_type_spec.rb]: Add a ticket?

File: spec/unit/pops/types/p_binary_type_spec.rb (hunk 12 above)
Comment (hlindberg) [inline spec/unit/pops/types/p_binary_type_spec.rb]: This should be replaced by adding the Binary type in the types_setup.rb - not meaningful to have another place to check assignability against all other types.
The test here (that you commented on) is quite meaningless. I am removing it.

File: lib/puppet/parser/functions/new.rb (hunk 14 above)
Comment (thallgren) [inline lib/puppet/parser/functions/new.rb]: I don't think that wording is correct either. The string does have an encoding, even if that encoding is unknown. It can't be ignored. Here, it's apparently up to Ruby to turn the string into a sequence of bytes. How does it do that?

Let's assume we would implement this in Java. A String in Java uses 16-bit characters. Would that mean that you'd get twice as many bytes using this method?

File: spec/unit/pops/types/p_binary_type_spec.rb (hunk 13 above)
Comment (thallgren) [inline spec/unit/pops/types/p_binary_type_spec.rb]: It's been added. See [PUP-6706](https://tickets.puppetlabs.com/browse/PUP-6706).

File: lib/puppet/parser/functions/new.rb [hunk 15]
```diff
@@ -864,7 +866,7 @@
 | b | The data is in base64 encoding, padding as required by base64 strict is added by default
 | u | The data is in URL safe base64 encoding
 | B | The data is in base64 strict encoding
-| s | The data is a puppet string used verbatim as a binary byte sequence
+| s | The data is a puppet string seen as a binary sequence of bytes with ignored encoding
```
Comment (thallgren) [inline lib/puppet/parser/functions/new.rb]: I don't think that wording is correct either. The string, when represented as bytes, does have an encoding, even if that encoding is unknown. It can't be ignored. Here, it's apparently up to Ruby to turn the string into a sequence of bytes. How does it do that?

Let's assume we would implement this in Java where a string uses 16-bit characters. Would that mean that you'd get twice as many bytes using this method?

Comment (hlindberg) [inline lib/puppet/parser/functions/new.rb]: Eh, it works the other way around. Ruby encodes when an operation takes place on the string. It is always a binary byte sequence internally. The ruby method that returns the bytes simply returns them. If you implemented this in Java, the Binary byte buffer created from a UTF-16 string would contain two bytes per character (or more since UTF-16 also has multi-chars IIRC). Do you have a suggestion on how to phrase this? Technically it takes a copy and enforces an '8 bit ascii' encoding (which is ruby's "no encoding, any byte sequence allowed" - i.e. "ignored encoding").

Comment (hlindberg) [inline lib/puppet/parser/functions/new.rb]: It needs to be phrased so that it is understandable by a puppet user without having to understand things like forced encoding - "uses the bytes of the string, whatever they are" to use a layman's term. Will change to "seen as the verbatim byte sequence representing the string in its given encoding" in anticipation you find that more correct.

Comment (thallgren) [inline lib/puppet/parser/functions/new.rb]: What I don't understand, is why we must expose the internal workings of one specific platform (Ruby) to the end user. That whole thing seems broken to me. Perhaps I don't understand the use-case, but the way I see it, there's just no way you can create a Binary from a string unless a decoding is given. The "use the bytes of the string, whatever they are" is a misnomer, since some string implementations aren't even based on bytes. The conversion opens up a can of worms as it exposes the internal runtime implementation in a bad way (to my knowledge, 8-bit-ascii doesn't exist outside of Ruby).

Can you describe how this is supposed to be used?

Comment (hlindberg) [inline lib/puppet/parser/functions/new.rb]: The use case is to enable use of Binary instead of just a String (as it is currently used), but where the encoding of the content is not known (it is just binary/raw data - it may be encoded in some text encoding, but that is not known). Ruby does not have a byte buffer, so currently functions return ascii 8 bit strings, or UTF-8 strings. I understand your concern regarding platform independence - would it help if it is specified as "8 bit ascii" (which I believe translates to ISO-8859-1 in Java as it allows any byte values in the string even if it implies that they represent particular glyphs). The alternative is to add the ability to specify the wanted encoding. Those are however platform specific. In any case, a port to some other language would need to revise what functions return - if it should return a Binary or a String and it would need to know when it is doing the reading. If you in Java were to do the stupid thing of creating a string that is really a binary, you could read it as ISO-8859-1.

Once the Binary type is introduced in Puppet we can deprecate / enforce (in some major version) that all puppet strings must be UTF-8, and that Binary is used for everything else.

Comment (hlindberg) [inline lib/puppet/parser/functions/new.rb]: The specification can state that for Ruby, the string will retain its encoding and used byte by byte. Other (future) platforms must create text with the correct encoding such that when mapped to a String, and then to a Binary using Binary.new(str, '%s') that it retains the read original byte sequence verbatim. It can also state that functions should not do this, and instead directly return Binary.

Comment (thallgren) [inline lib/puppet/parser/functions/new.rb]: "8-bit-ascii" does not help, since its just as platform dependent as saying nothing.

When converting a Binary to a String using '%s', the encoding will be "UTF-8". The most consistent approach would be to then say that 's' here, also means "UTF-8". If not, then converting to and through using '%s' and 's' will give somewhat surprising results.

My preference would therefore be to say that the default encoding is "UTF-8" but that other encodings can be specified. We must then either not allow "8-bit-ascii" since it doesn't exist outside the Ruby community or allow it but raise a red flag stating that it is 100% platform specific and should be avoided at length. My preference would be the former and instead provide a set of encodings known to be standardized.

Comment (hlindberg) [inline lib/puppet/parser/functions/new.rb]: There must be a way to get "the encoding that is used by the String", which in Ruby means "whatever" - it is not in our power to change that right here. We also do not have any other platforms to consider atm. When we do, that implementation must be done such that reading a binary byte sequence into a String (a totally stupid thing to do that should be forbidden as Binary should be used, but if you went ahead and did that anyway) it must either use the same style as Ruby and embed the encoding in the string, or be read so that Binary(str, '%s') produces the same byte sequence as was originally read (details how it does that is of no importance but could be reading ISO-8859-1, and then converting with the same encoding).

Comment (hlindberg) [inline lib/puppet/parser/functions/new.rb]: In puppet language the distinction between how something is read is done by calling `file >> String` (assumed but not enforced UTF-8), or `binary_file >> Binary` (raw, bytes). We have however no control whatsoever over functions in the wild, and there is no specification other than the existing behavior where puppet is using the encoding embedded in the Ruby string. Note that the `file` read does not enforce UTF-8 - as it performs a binary read and any later requirement that the format must be UTF-8 makes things blow up later.

Comment (thallgren) [inline lib/puppet/parser/functions/new.rb]: Do you agree that '%s' should use the same encoding when converting to/from a Binary? If the answer to that is yes, then "%s" has to mean "UTF-8". Some other format must then be added to mean "whatever" or, perhaps "ISO-8859-1". As you mentioned before, it's a valid encoding that will accept the full 8-bit range of bytes. I wouldn't be surprised if it's a no-op in Ruby (i.e. bytes == string characters).

**Comment by hlindberg** [comment]
THe %s for Binary.new is now changed to require valid UTF-8, or string transcodable to valid UTF-8. A new format %r is added to do what %s did earlier. That solution is symmetrical (that is `String.new(bin, "%s")` and `Binary.new(str, %s)` both operate on UTF-8 strings). The `%r` format is specified as a backwards compatibility temporary solution that eventually will be deprecated.
